File: src/NodeParser/NamedTupleMemberNodeParser.ts
import ts from "typescript";
import { Context, NodeParser } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { AnnotatedType } from "../Type/AnnotatedType.js";
import { ArrayType } from "../Type/ArrayType.js";
import { BaseType } from "../Type/BaseType.js";
import { ReferenceType } from "../Type/ReferenceType.js";
import { RestType } from "../Type/RestType.js";

export class NamedTupleMemberNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.TypeNode): boolean {
        return node.kind === ts.SyntaxKind.NamedTupleMember;
    }

    public createType(node: ts.NamedTupleMember, context: Context, reference?: ReferenceType): BaseType {
        const baseType = this.childNodeParser.createType(node.type, context, reference);

        if (baseType instanceof ArrayType && node.getChildAt(0).kind === ts.SyntaxKind.DotDotDotToken) {
            return new RestType(baseType, node.name.text);
        }

        return baseType && new AnnotatedType(baseType, { title: node.name.text }, false);
    }
}

File: internal/parser/named_tuple_member_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// NamedTupleMemberNodeParser handles named tuple members like
// `[name: string]` (src/NodeParser/NamedTupleMemberNodeParser.ts).
type NamedTupleMemberNodeParser struct {
	childNodeParser NodeParser
}

func NewNamedTupleMemberNodeParser(childNodeParser NodeParser) *NamedTupleMemberNodeParser {
	return &NamedTupleMemberNodeParser{childNodeParser: childNodeParser}
}

func (p *NamedTupleMemberNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindNamedTupleMember
}

func (p *NamedTupleMemberNodeParser) CreateType(node *ast.Node, ctx *Context, reference *types.ReferenceType) types.Type {
	member := node.AsNamedTupleMember()
	baseType := p.childNodeParser.CreateType(member.Type, ctx, reference)

	if _, isArray := baseType.(*types.ArrayType); isArray && member.DotDotDotToken != nil {
		return &types.RestType{Type: baseType, Title: member.Name().Text()}
	}

	if baseType == nil {
		return nil
	}
	return &types.AnnotatedType{
		Type:        baseType,
		Annotations: types.Annotations{"title": member.Name().Text()},
		Nullable:    false,
	}
}
